Lib/fractions.py

```diff
@@ -382,30 +382,59 @@ def reverse(b, a):
 
     def _add(a, b):
         """a + b"""
+        na, nb = a.numerator, b.numerator
         da, db = a.denominator, b.denominator
-        return Fraction(a.numerator * db + b.numerator * da,
-                        da * db)
+        g = math.gcd(da, db)
+        if g == 1:
+            # This is the case with probability 6/pi**2 ~ 60.8%.
+            return Fraction(na * db + da * nb, da * db, _normalize=False)
+        else:
+            q1, q2 = da // g, db // g
+            n, d = na * q2 + nb * q1, q1 * q2
+            g2 = math.gcd(n, g)
+            return Fraction(n // g2, d * (g // g2), _normalize=False)
 
     __add__, __radd__ = _operator_fallbacks(_add, operator.add)
 
     def _sub(a, b):
         """a - b"""
+        na, nb = a.numerator, b.numerator
         da, db = a.denominator, b.denominator
-        return Fraction(a.numerator * db - b.numerator * da,
-                        da * db)
+        g = math.gcd(da, db)
+        if g == 1:
+            # This is the case with probability 6/pi**2 ~ 60.8%.
+            return Fraction(na * db - da * nb, da * db, _normalize=False)
+        else:
+            q1, q2 = da // g, db // g
+            n, d = na * q2 - nb * q1, q1 * q2
+            g2 = math.gcd(n, g)
+            return Fraction(n // g2, d * (g // g2), _normalize=False)
 
     __sub__, __rsub__ = _operator_fallbacks(_sub, operator.sub)
 
     def _mul(a, b):
         """a * b"""
-        return Fraction(a.numerator * b.numerator, a.denominator * b.denominator)
+        na, da = a.numerator, a.denominator
+        nb, db = b.numerator, b.denominator
+        x1 = math.gcd(na, db)
+        x2 = math.gcd(nb, da)
+        n, d = (na//x1)*(nb//x2), (da//x2)*(db//x1)
+        if d < 0:
```

Comment:
Is there a way that *_mul()* gets called with unnormalized inputs?  If not, then this test could be skipped since the *gcd()* outputs are non-negative.

Reply:
No, it's not.  But `_mul()` and `_div()` do unnormalized outputs (in sense of denominator signs), see e.g. [this](https://github.com/python/cpython/runs/2049106689) test failure.  I suspect similar for `_add()/_sub()` (I haven't proved reversed so far), but the test suite doesn't have such cases.

Reply:
Oh, no.  For `_add()/_sub()` - both `q1` and `q2` are positive for normalized inputs and `n` is positive too (as `gcd` output).  Fixing the denominator sign is not required here.

Reply:
Aren't _all_ of `da`, `db`, `x1` and `x2` strictly positive? Can you give an example of an operation that would exercise this branch? I can see why we need to make this check for division (and it looks as though the test failure you link to relates to division rather than multiplication), but I can't see how this branch would be needed.

Reply:
Indeed, only `_div()` output should be normalized (here negative sign may come from `nb`).